api/envoy/extensions/filters/http/credential_injector/v3alpha/credential_injector.proto

```diff
@@ -0,0 +1,111 @@
+syntax = "proto3";
+
+package envoy.extensions.filters.http.credentials.v3alpha;
+
+import "envoy/config/core/v3/http_uri.proto";
+import "envoy/extensions/transport_sockets/tls/v3/secret.proto";
+
+import "udpa/annotations/status.proto";
+import "validate/validate.proto";
+
+option java_package = "io.envoyproxy.envoy.extensions.filters.http.credential_injector.v3alpha";
+option java_outer_classname = "CredentialInjectorProto";
+option java_multiple_files = true;
+option go_package = "github.com/envoyproxy/go-control-plane/envoy/extensions/filters/http/credential_injector/v3alpha;credential_injectorv3alpha";
+option (udpa.annotations.file_status).package_version_status = ACTIVE;
+
+// [#protodoc-title: Credentials Injector]
+// Credentials Injector :ref:`configuration overview <config_http_filters_credential_injector>`.
+// [#extension: envoy.filters.http.credential_injector]
+//
+
+// Filter config.
+message CredentialInjector {
+  // Credential to inject into proxied requests.
+  Credential credential = 1 [(validate.rules).message = {required: true}];
+}
+
+// Credential to inject into proxied requests.
+//
+// [#next-free-field: 5]
+message Credential {
+  oneof credential_type {
+    option (validate.required) = true;
+
+    // Basic Auth Credential to inject.
+    BasicAuthCredential basic = 1;
+
+    // Bearer Token Credential to inject.
+    transport_sockets.tls.v3.SdsSecretConfig bearer = 2;
+
+    // OAuth2 Credential to inject.
+    OAuth2Credential oauth2 = 3;
+  }
+}
+
+// Basic Auth Credential to inject into proxied requests.
+//
+// RFC https://www.rfc-editor.org/rfc/rfc7617
+//
+// [#next-free-field: 3]
+message BasicAuthCredential {
+  // User name.
+  string username = 1 [(validate.rules).string = {min_len: 1}];
+
+  // User password.
+  transport_sockets.tls.v3.SdsSecretConfig password = 2
+      [(validate.rules).message = {required: true}];
+}
+
+// OAuth2 Credential to inject into proxied requests.
+//
+// [#next-free-field: 4]
+message OAuth2Credential {
+  // Endpoint on the authorization server to retrieve the access token from.
+  // OAuth RFC https://www.rfc-editor.org/rfc/rfc6749#section-3.2
+  config.core.v3.HttpUri token_endpoint = 1 [(validate.rules).message = {required: true}];
+
+  // Optional list of OAuth scopes to be claimed in the authorization request.
+  // OAuth RFC https://www.rfc-editor.org/rfc/rfc6749#section-4.4.2
+  repeated string scopes = 2;
+
+  enum AuthType{
+    // The ``client_id`` and ``client_secret`` will be sent using HTTP Basic authentication scheme.
+    BASIC_AUTH = 0;
+
+    // The ``client_id`` and ``client_secret`` will be sent in the URL encoded request body.
+    // This type should only be used when Auth server does not support Basic authentication.
+    URL_ENCODED_BODY = 1;
```

Comment:
I decide to keep it this way because the first element in an Enum is the default value, and according to oauth spec, the Basic authentication scheme is used by default.